### PR TITLE
EPMRPP-37610 Implemented most time consuming test cases widget

### PR DESCRIPTION
**Pull request by twinfreaks**

### CODE REVIEW

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCases.jsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import { MostTimeConsumingTestCasesChart } from './mostTimeConsumingTestCasesChart';
+import { MostTimeConsumingTestCasesTable } from './mostTimeConsumingTestCasesTable';
+import styles from './mostTimeConsumingTestCases.scss';
+
+const cx = classNames.bind(styles);
+
+const localMessages = defineMessages({
+  launchNameText: {
+    id: 'TimeConsuming.launchNameText',
+    defaultMessage: 'LAUNCH NAME:',
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCases.jsx]: Please write it as usual case. Uppercase should be applied by CSS rule

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    };
+
+    this.size = params.size;
+
+    const configParams = {
+      ...params,
+      chartType: MODES_VALUES[CHART_MODES.BAR_VIEW],
+      isPointsShow: false,
+      messages: MESSAGES,
+      isCustomTooltip: true,
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx]: Unused

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/getConfig.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  chartType = MODES_VALUES[CHART_MODES.BAR_VIEW],
+  isPointsShow = true,
+  testCaseClickHandler,
+}) => {
+  const data = [...content.result]
+    .sort((a, b) => b.duration - a.duration)
+    .map((item, index) => ({ ...item, tickIndex: 30 - index, index }));
+
+  const getItemColor = (color, { index }) =>
+    index % 2 ? COLOR_TIMECONSUMING_BAR_GREEN : COLOR_TIMECONSUMING_BAR_RED;
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/getConfig.js]: Color should be calculated based on item status, not index

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/getConfig.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import { PERIOD_VALUES } from 'common/constants/statusPeriodValues';
+import {
+  COLOR_TIMECONSUMING_BAR_RED,
+  COLOR_TIMECONSUMING_BAR_GREEN,
+} from 'common/constants/colors';
+import { CHART_MODES, MODES_VALUES } from 'common/constants/chartModes';
+import { MostTimeConsumingTooltip } from '../mostTimeConsumingTooltip/index';
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/getConfig.js]: `import { MostTimeConsumingTooltip } from '../mostTimeConsumingTooltip';`

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames/bind';
+import { injectIntl, intlShape } from 'react-intl';
+import { TEST_ITEM_PAGE, PROJECT_LOG_PAGE } from 'controllers/pages/constants';
+import { CHART_MODES, MODES_VALUES } from 'common/constants/chartModes';
+import { ALL } from 'common/constants/reservedFilterIds';
+import { C3Chart } from '../../charts/common/c3chart/index';
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx]: `import { C3Chart } from 'components/widgets/charts/common/c3chart';`

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    isConfigReady: false,
+  };
+
+  componentDidMount() {
+    const { observer } = this.props;
+
+    observer.subscribe && observer.subscribe('widgetResized', this.resizeChart);
+    this.getConfig();
+  }
+
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx]: Please also provide
```
 componentDidUpdate(prevProps) {
    if (!isEqual(prevProps.widget, this.props.widget)) {
      this.getConfig();
    }
  }
```
to correct widget update

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/getConfig.js (hunk 3 above)
Comment (twinfreaks) [reply app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/getConfig.js]: There are 2 stripe colors alternating each other - green and red, not depending on status

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCases.jsx [hunk 7]
```diff
@@ -16,7 +16,7 @@ const cx = classNames.bind(styles);
 const localMessages = defineMessages({
   launchNameText: {
     id: 'TimeConsuming.launchNameText',
-    defaultMessage: 'LAUNCH NAME:',
+    defaultMessage: 'Launch name:',
```
Comment (AmsterGet) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCases.jsx]: Please add `text-transform: uppercase` css property to `launch-name-text` class

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  testCaseClickHandler = (id) => {
+    const {
+      project,
+      widget: {
+        content: { result, latestLaunch = {} },
+      },
+      navigate,
+    } = this.props;
+    const targetElement = result.filter((el) => el.id === id)[0] || {};
```
Comment (DiscoElevator) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTestCasesChart/mostTimeConsumingTestCasesChart.jsx]: `result.find`

File: app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTooltip/mostTimeConsumingTooltip.jsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import ReactDOMServer from 'react-dom/server';
+import Parser from 'html-react-parser';
+import { DATETIME_FORMAT_TOOLTIP } from 'common/constants/timeDateFormat';
+import TimeIcon from 'common/img/time-icon-inline.svg';
+import { TooltipWrapper } from '../../charts/common/tooltip/tooltipWrapper/index';
+import styles from './mostTimeConsumingTooltip.scss';
+
+const cx = classNames.bind(styles);
+
+export const MostTimeConsumingTooltip = (data) => (d) => {
```
Comment (DiscoElevator) [inline app/src/components/widgets/mostTimeConsumingTestCases/mostTimeConsumingTooltip/mostTimeConsumingTooltip.jsx]: This lools like a component, but it's not. In this case it's a factory that creates function which renders static markup.
I'd better split this into a tooltip component and a render function which renders this tooltip to a static markup.
Or just renamed this to something like `createMostTimeConsumingTooltipRenderer`.

**Pull request merged**